src/materialized/tests/sql.rs

```diff
@@ -943,3 +943,41 @@ write frontier:[            0]\n",
 
     Ok(())
 }
+
+// Test that a query that causes a compute instance to panic will resolve
+// the panic and allow the compute instance to restart (instead of crash loop
+// forever) when a client is terminated (disconnects from the server) instead
+// of cancelled (sends a pgwire cancel request on a new connection).
+#[test]
+fn test_github_12546() -> Result<(), Box<dyn Error>> {
+    mz_ore::test::init_logging();
+    let config = util::Config::default();
+    let server = util::start_server(config)?;
+
+    server.runtime.block_on(async {
+        let (client, conn_task) = server.connect_async(tokio_postgres::NoTls).await?;
+
+        client.batch_execute("CREATE TABLE test(a text);").await?;
+        client
+            .batch_execute("INSERT INTO test VALUES ('a');")
+            .await?;
+
+        let query = client.query("SELECT mz_internal.mz_panic(a) FROM test", &[]);
+        let timeout = tokio::time::timeout(Duration::from_secs(2), query);
+        // We expect the timeout to trigger because the query should be crashing the
+        // compute instance.
+        assert_eq!(
+            timeout.await.unwrap_err().to_string(),
+            "deadline has elapsed"
+        );
```

Comment:
I'm confused here because this seems like it just guarantees that the query didn't return in 2 seconds--I can't follow the thread that shows where the compute instance successfully restarts, as mentioned in the test comment. If it's too complicated to type out, would love to hop on a call.

Reply:
Yeah—is there a missing step after aborting the original task where you verify that new queries against the cluster now work?

Reply:
The test as is does fail if the connection drop detection is removed, so *something* is waiting for this to happen. But I have no idea what.

Did some checking and it was like the end of test Drop or something. Dunno. Added a thing that verifies it can query.

Reply:
Sweet, yeah, that's much clearer!